HoRNDIS.cpp

```diff
@@ -28,516 +30,343 @@
+void HoRNDIS::closeUSBInterfaces() {
+	fReadyToTransfer = false;  // Interfaces are about to be closed.
+	// Close the interfaces - this would abort the transfers (if present):
+	if (fDataInterface) {
+		fDataInterface->close(this);
 	}
-	
-	return false;
+	if (fCommInterface) {
+		fCommInterface->close(this);
+	}
+
+	OSSafeReleaseNULL(fInPipe);
+	OSSafeReleaseNULL(fOutPipe);
+	OSSafeReleaseNULL(fDataInterface);
+	OSSafeReleaseNULL(fCommInterface);  // First one to open, last one to die.
 }
 
 IOService *HoRNDIS::probe(IOService *provider, SInt32 *score) {
-	LOG(V_NOTE, "probe: came in with a score of %d", *score);
-	IOUSBDevice *dev = OSDynamicCast(IOUSBDevice, provider); /* XXX: error check */
-	
-	/* Do we even have one of the things that we can match on?  If not, return NULL.  Either 2/255/2, or 224/3/1.  */
-	const IOUSBConfigurationDescriptor *cd;
-	
-	cd = dev->GetFullConfigurationDescriptor(0);
-	if (!cd) {
-		LOG(V_ERROR, "probe: failed to get a configuration descriptor for configuration 0?");
+	LOG(V_DEBUG, "came in with a score of %d", *score);
+	
+	// Driver matching algorithm: keep it simple for now. Assumptions:
+	//  - Android device has only one USB configuration OR RNDIS interfaces
+	//    would be in the active configuration.
+	// Algorithm:
+	//  - The Info.plist matchies on "224/1/3" - RNDIS control interface.
+	//  - "probe" double-checks that and also makes sure the data interface is present.
+	//  - The "start" can open the required interfaces right away - no need to open
+	//    the device and set the configuration.
+	IOUSBHostInterface *interface = OSDynamicCast(IOUSBHostInterface, provider);
+	if (interface == NULL) {
+		LOG(V_ERROR, "unexpected provider class (wrong Info.plist)");
 		return NULL;
 	}
 	
-	/* And look for one of the options. */
-	IOUSBInterfaceDescriptor *descout;
-	IOUSBFindInterfaceRequest req;
-	bool found = false;
-	
-	ctrlconfig = cd->bConfigurationValue;
-	
-	req.bInterfaceClass = 2;
-	req.bInterfaceSubClass = 2;
-	req.bInterfaceProtocol = 255;
-	if (dev->FindNextInterfaceDescriptor(cd, NULL, &req, &descout) == kIOReturnSuccess) {
-		LOG(V_NOTE, "probe: looks like we're good (2/2/255)");
-		ctrlclass = 2;
-		ctrlsubclass = 2;
-		ctrlprotocol = 255;
-		found = true;
+	if (!isRNDISControlInterface(interface->getInterfaceDescriptor())) {
+		LOG(V_ERROR, "not RNDIS control interface (wrong Info.plist)");
+		return NULL;
 	}
-
-	req.bInterfaceClass = 224;
-	req.bInterfaceSubClass = 1;
-	req.bInterfaceProtocol = 3;
-	if (dev->FindNextInterfaceDescriptor(cd, NULL, &req, &descout) == kIOReturnSuccess) {
-		ctrlclass = 224;
-		ctrlsubclass = 1;
-		ctrlprotocol = 3;
-		LOG(V_NOTE, "probe: looks like we're good (224/1/3)");
-		found = true;
+	const uint8_t controlIfNum = interface->getInterfaceDescriptor()->bInterfaceNumber;
+	
+	// Now, search for data interface: if found, we're done:
```

Comment:
Oh, I see, this answers the question posed above.

Reply:
ack